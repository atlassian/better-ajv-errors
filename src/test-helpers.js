import { readFileSync } from 'fs';
import { getFixturePath } from 'jest-fixtures';
import Ajv from 'ajv';

export async function getSchemaAndData(name, dirPath) {
  const schemaPath = await getFixturePath(dirPath, name, 'schema.json');
  const schema = JSON.parse(readFileSync(schemaPath, 'utf8'));
  const dataPath = await getFixturePath(dirPath, name, 'data.json');
  const data = JSON.parse(readFileSync(dataPath, 'utf8'));

  return [schema, data];
}

export function evaluateSchema(schema, data) {
  const ajv = new Ajv();
  const fn = ajv.compile(schema);
  fn(data);
  expect(fn.errors.length).toBeGreaterThanOrEqual(0);
  return fn.errors;
}
