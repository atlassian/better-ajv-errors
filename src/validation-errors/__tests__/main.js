import Ajv from 'ajv';
import betterAjvErrors from '../../';
import { getSchemaAndData } from '../../test-helpers';

describe('Main', () => {
  it('should support js output format for default errors', async () => {
    const [schema, data] = await getSchemaAndData('default', __dirname);

    const ajv = new Ajv({ jsonPointers: true });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBeFalsy();

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for required errors', async () => {
    const [schema, data] = await getSchemaAndData('required', __dirname);

    const ajv = new Ajv({ jsonPointers: true });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBeFalsy();

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for additionalProperties errors', async () => {
    const [schema, data] = await getSchemaAndData(
      'additionalProperties',
      __dirname
    );

    const ajv = new Ajv({ jsonPointers: true });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBeFalsy();

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });

  it('should support js output format for enum errors', async () => {
    const [schema, data] = await getSchemaAndData('enum', __dirname);

    const ajv = new Ajv({ jsonPointers: true });
    const validate = ajv.compile(schema);
    const valid = validate(data);
    expect(valid).toBeFalsy();

    const res = betterAjvErrors(schema, data, validate.errors, {
      format: 'js',
    });
    expect(res).toMatchSnapshot();
  });
});
