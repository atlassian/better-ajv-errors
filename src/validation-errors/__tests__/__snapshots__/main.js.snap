// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main should support js output format for additionalProperties errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 27,
      "line": 1,
      "offset": 26,
    },
    "error": "Property baz is not expected to be here",
    "start": Object {
      "column": 22,
      "line": 1,
      "offset": 21,
    },
  },
]
`;

exports[`Main should support js output format for default errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 25,
      "line": 1,
      "offset": 24,
    },
    "error": "/id: type should be number",
    "start": Object {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
  },
]
`;

exports[`Main should support js output format for enum errors 1`] = `
Array [
  Object {
    "end": Object {
      "column": 12,
      "line": 1,
      "offset": 11,
    },
    "error": "/id should be equal to one of the allowed values: foo, bar",
    "start": Object {
      "column": 7,
      "line": 1,
      "offset": 6,
    },
    "suggestion": "Did you mean bar?",
  },
]
`;

exports[`Main should support js output format for required errors 1`] = `
Array [
  Object {
    "error": "Required property id is missing",
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
]
`;
